lib/jerboa/client/worker.ex

```diff
@@ -4,70 +4,166 @@ defmodule Jerboa.Client.Worker do
+    lifetime: non_neg_integer,
+    lifetime_timer_ref: reference
+  }
 
   @system_allocated_port 0
 
-  def start_link(x) do
-    GenServer.start_link(__MODULE__, x)
+  @spec start_link(Client.start_opts) :: GenServer.on_start
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
   end
 
-  def init(address: a, port: p) do
+  ## GenServer callbacks
+
+  def init(opts) do
     false = Process.flag(:trap_exit, true)
     {:ok, socket} = UDP.open(@system_allocated_port, [:binary, active: false])
-    {:ok,
-     %__MODULE__{server: {a, p}, socket: socket}}
+    state = Protocol.init_state(opts, socket)
+    {:ok, state}
   end
 
   def handle_call(:bind, _, state) do
-    msg = Jerboa.Format.encode(Jerboa.Params.put_class(binding_(), :request))
-    response = call(socket(state), server(state), msg)
-    {:ok, params} = Jerboa.Format.decode(response)
-    {:reply, reflexive_candidate(params), state}
+    {result, new_state} =
+      state
+      |> Protocol.bind_req()
+      |> send_req()
+      |> Protocol.eval_bind_resp()
+    {:reply, result, new_state}
+  end
+  def handle_call(:allocate, _, state) do
+    if state.relayed_address do
+      {:reply, {:ok, state.relayed_address}, state}
+    else
+      {result, new_state} = request_allocation(state)
+      case result do
+        {:ok, _} ->
+          {:reply, result, update_lifetime_timer(new_state)}
+        _ ->
+          {:reply, result, new_state}
+      end
+    end
+  end
+  def handle_call(:refresh, _, state) do
+    unless state.relayed_address do
+      {:reply, {:error, :no_allocation}, state}
+    else
+      {result, new_state} = request_refresh(state)
+      case result do
+        :ok ->
+          {:reply, result, update_lifetime_timer(new_state)}
+        _ ->
+          {:reply, result, new_state}
+      end
+    end
   end
-  def handle_call(:persist, _, state) do
-    msg = Jerboa.Format.encode(Jerboa.Params.put_class(binding_(), :indication))
-    {:reply, cast(socket(state), server(state), msg), state}
+
+  def handle_cast(:persist, state) do
+    state
+    |> Protocol.bind_ind()
+    |> send_ind()
+    {:noreply, state}
+  end
+
+  def handle_info(:allocation_expired, state) do
+    new_state = %{state | lifetime_timer_ref: nil,
+                          relayed_address: nil,
+                          lifetime: nil}
+    {:noreply, new_state}
   end
 
   def terminate(_, state) do
     :ok = UDP.close(socket(state))
   end
 
-  defp server(%__MODULE__{server: s}) do
-    s
-  end
+  ## Internals
+
+  defp server(%{server: server}), do: server
 
-  defp socket(%__MODULE__{socket: s}) do
-    s
+  defp socket(%{socket: socket}), do: socket
+
+  @spec send_req(state) :: state
+  defp send_req(%{transaction: %{req: req} = transaction} = state) do
+    socket = socket(state)
+    {address, port} = server(state)
+    :ok = UDP.send(socket, address, port, req)
+    {:ok, {^address, ^port, response}} = UDP.recv(socket, 0, timeout())
```

Comment:
Everytime you need timeout you take it our from the App env? no no no, that's not that good idea.. why not putting it into the server's state?